controllers/space/space_controller.go

```diff
@@ -191,68 +199,96 @@ func (r *Reconciler) ensureNSTemplateSet(logger logr.Logger, space *toolchainv1a
 		if errors.IsNotFound(err) {
 			logger.Info("creating NSTemplateSet on target member cluster")
 			if err := r.setStatusProvisioning(space); err != nil {
-				return false, r.setStatusProvisioningFailed(logger, space, err)
+				return norequeue, r.setStatusProvisioningFailed(logger, space, err)
 			}
 			nsTmplSet = r.newNSTemplateSet(memberCluster.OperatorNamespace, space.Name, tmplTier)
 
 			if err := memberCluster.Client.Create(context.TODO(), nsTmplSet); err != nil {
 				logger.Error(err, "failed to create NSTemplateSet on target member cluster")
-				return false, r.setStatusNSTemplateSetCreationFailed(logger, space, err)
+				return norequeue, r.setStatusNSTemplateSetCreationFailed(logger, space, err)
 			}
 			logger.Info("NSTemplateSet created on target member cluster")
-			return true, r.setStatusProvisioning(space)
+			return requeueDelay, r.setStatusProvisioning(space)
 		}
-		return false, r.setStatusNSTemplateSetCreationFailed(logger, space, err)
+		return norequeue, r.setStatusNSTemplateSetCreationFailed(logger, space, err)
 	}
 	logger.Info("NSTemplateSet already exists")
 
-	// update the NSTemplateSet if needed
-	if !tierutil.TierHashMatches(tmplTier, nsTmplSet.Spec) {
+	spaceReady, found := condition.FindConditionByType(space.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	// skip until there's a `Ready` condition
+	if !found {
+		// just created, but there is no `Ready` condition yet
+		return requeueDelay, nil
+	}
+	nsTmplSetReady, found := condition.FindConditionByType(nsTmplSet.Status.Conditions, toolchainv1alpha1.ConditionReady)
+	// skip until there's a `Ready` condition
+	if !found {
+		// just created, but there is no `Ready` condition yet
+		return requeueDelay, nil
+	}
+
+	tiersMatch := tierutil.TierHashMatches(tmplTier, nsTmplSet.Spec)
+
+	// update the NSTemplateSet if needed and if it's "ready"
+	if !tiersMatch &&
+		spaceReady.Reason == toolchainv1alpha1.SpaceProvisionedReason &&
+		nsTmplSetReady.Reason == toolchainv1alpha1.NSTemplateSetProvisionedReason {
```

Comment:
why do we care about the states of the conditions of the `Space` and `NSTemplateSet` here? If either the Space CR or the NSTemplateSet CR doesn't match the desired state, then I would expect that the controller should try to achieve the state no matter what kind of condition the CRs have. 

Just thought - NSTemplateSet can be in a failed state because they are not up-to-date and this if check would block from updating and thus also fixing them:
1. We modify the templates, but we introduce a bug
2. We merge it and it starts updating the Namespaces
3. Because of the bug, the NSTemplateSet ends up in a failed state
4. We fix the template and merge the PR
5. The NSTemplateSet never gets fixed because of this check

Reply:
the tier hashes will not match until the update is complete. So, without checking the state of the space and the nstemplateset, the controller will enter this "if" block on every reconcile loop that is triggered during the update. I don't think it's something we want.

Reply:
however, I don't see what a change in space bindings should trigger an update of the NSTemplateSet itself?

Reply:
> When there is a change in any SpaceBinding, then we need to reflect the change also in NSTemplateSet.Spec

Could you describe which change needs to be reflected in the NSTemplateSet?

Reply:
the SpaceRoles defined in the SpaceBindings and the associated usernames https://github.com/codeready-toolchain/api/blob/master/api/v1alpha1/nstemplateset_types.go#L75-L82
to make this functionality working https://github.com/codeready-toolchain/member-operator/commit/fc7708c3cfc1e75b9acca49e6fe0fda3db8150aa

Reply:
ah right! I totally forgot that we will also keep track of the usernames in the NSTemplateSet! 🤦‍♂️
Ok, so I will make compare the specs instead of the hashes, then.